testing-common/src/main/java/org/hypertrace/agent/testing/AbstractHttpClientTest.java

```diff
@@ -0,0 +1,288 @@
+  }
+
+  @Test
+  public void getNoContent()
+      throws IOException, TimeoutException, InterruptedException, ExecutionException {
+    String uri = String.format(GET_NO_CONTENT_PATH_FORMAT, testHttpServer.port());
+
+    Response response = doGetRequest(uri, headers);
+
+    Assertions.assertEquals(204, response.statusCode);
+    Assertions.assertNull(response.body);
+
+    TEST_WRITER.waitForTraces(1);
+    List<List<SpanData>> traces = TEST_WRITER.getTraces();
+    Assertions.assertEquals(1, traces.size());
+    Assertions.assertEquals(1, traces.get(0).size());
+    SpanData clientSpan = traces.get(0).get(0);
+
+    assertNoBodies(clientSpan);
+  }
+
+  @Test
+  public void getJson()
+      throws IOException, TimeoutException, InterruptedException, ExecutionException {
+    String uri = String.format(GET_JSON_PATH_FORMAT, testHttpServer.port());
+
+    Response response = doGetRequest(uri, headers);
+
+    Assertions.assertEquals(200, response.statusCode);
+    Assertions.assertEquals(TestHttpServer.GetJsonHandler.RESPONSE_BODY, response.body);
+
+    TEST_WRITER.waitForTraces(1);
+    if (hasResponseBodySpan) {
+      TEST_WRITER.waitForSpans(2);
+    }
+    List<List<SpanData>> traces = TEST_WRITER.getTraces();
+    Assertions.assertEquals(1, traces.size());
+    SpanData clientSpan = traces.get(0).get(0);
+    Assertions.assertNull(
+        clientSpan.getAttributes().get(HypertraceSemanticAttributes.HTTP_REQUEST_BODY));
+
+    if (hasResponseBodySpan) {
+      Assertions.assertEquals(2, traces.get(0).size());
+      SpanData responseBodySpan = traces.get(0).get(1);
+      Assertions.assertEquals(
+          TestHttpServer.GetJsonHandler.RESPONSE_BODY,
+          responseBodySpan.getAttributes().get(HypertraceSemanticAttributes.HTTP_RESPONSE_BODY));
+    } else {
+      Assertions.assertEquals(1, traces.get(0).size());
+      Assertions.assertEquals(
+          TestHttpServer.GetJsonHandler.RESPONSE_BODY,
+          clientSpan.getAttributes().get(HypertraceSemanticAttributes.HTTP_RESPONSE_BODY));
+    }
+  }
+
+  private void assertHeaders(SpanData spanData) {
+    Assertions.assertEquals(
+        TestHttpServer.RESPONSE_HEADER_VALUE,
+        spanData
+            .getAttributes()
+            .get(
+                HypertraceSemanticAttributes.httpResponseHeader(
+                    TestHttpServer.RESPONSE_HEADER_NAME)));
+    Assertions.assertEquals(
+        HEADER_VALUE,
+        spanData.getAttributes().get(HypertraceSemanticAttributes.httpRequestHeader(HEADER_NAME)));
+  }
+
+  private void assertEchoBodyInSpan(SpanData spanData, String requestBody) {
+    Assertions.assertEquals(
+        requestBody, spanData.getAttributes().get(HypertraceSemanticAttributes.HTTP_REQUEST_BODY));
+    Assertions.assertEquals(
+        spanData.getAttributes().get(HypertraceSemanticAttributes.HTTP_RESPONSE_BODY),
+        spanData.getAttributes().get(HypertraceSemanticAttributes.HTTP_REQUEST_BODY));
+  }
+
+  private void assertEchoBodyInSpans(
```

Comment:
change this to `assertBodies(RequestSpan, ResponseSpan, requestBody, ResponseBody)